In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
store = pd.read_csv("..data/store.csv")

In [3]:
store.head()

,Store,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,c,a,1270.0,9.0,2008.0,0,NaN,NaN,NaN
1,2,a,a,570.0,11.0,2007.0,1,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,a,a,14130.0,12.0,2006.0,1,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,c,c,620.0,9.0,2009.0,0,NaN,NaN,NaN
4,5,a,a,29910.0,4.0,2015.0,0,NaN,NaN,NaN


In [5]:
store.isnull().sum()

Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            3
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64

In [6]:
store.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1115 entries, 0 to 1114
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Store                      1115 non-null   int64  
 1   StoreType                  1115 non-null   object 
 2   Assortment                 1115 non-null   object 
 3   CompetitionDistance        1112 non-null   float64
 4   CompetitionOpenSinceMonth  761 non-null    float64
 5   CompetitionOpenSinceYear   761 non-null    float64
 6   Promo2                     1115 non-null   int64  
 7   Promo2SinceWeek            571 non-null    float64
 8   Promo2SinceYear            571 non-null    float64
 9   PromoInterval              571 non-null    object 
dtypes: float64(5), int64(2), object(3)
memory usage: 87.2+ KB


In [7]:
train = pd.read_csv("C:\\Users\\medha\\OneDrive\\Desktop\\Retail Promotion Effectiveness Analyzer Project\\train.csv")

C:\Users\medha\AppData\Local\Temp\ipykernel_9668\2685213707.py:1: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  train = pd.read_csv("C:\\Users\\medha\\OneDrive\\Desktop\\Retail Promotion Effectiveness Analyzer Project\\train.csv")


In [8]:
train.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday
0,1,5,2015-07-31,5263,555,1,1,0,1
1,2,5,2015-07-31,6064,625,1,1,0,1
2,3,5,2015-07-31,8314,821,1,1,0,1
3,4,5,2015-07-31,13995,1498,1,1,0,1
4,5,5,2015-07-31,4822,559,1,1,0,1


In [9]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB


In [10]:
train.shape

(1017209, 9)

In [11]:
store.shape

(1115, 10)

In [12]:
df = train.merge(store, on="Store", how="left")

In [13]:
df.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,5,2015-07-31,5263,555,1,1,0,1,c,a,1270.0,9.0,2008.0,0,NaN,NaN,NaN
1,2,5,2015-07-31,6064,625,1,1,0,1,a,a,570.0,11.0,2007.0,1,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,5,2015-07-31,8314,821,1,1,0,1,a,a,14130.0,12.0,2006.0,1,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,5,2015-07-31,13995,1498,1,1,0,1,c,c,620.0,9.0,2009.0,0,NaN,NaN,NaN
4,5,5,2015-07-31,4822,559,1,1,0,1,a,a,29910.0,4.0,2015.0,0,NaN,NaN,NaN


In [14]:
df.columns

Index(['Store', 'DayOfWeek', 'Date', 'Sales', 'Customers', 'Open', 'Promo',
       'StateHoliday', 'SchoolHoliday', 'StoreType', 'Assortment',
       'CompetitionDistance', 'CompetitionOpenSinceMonth',
       'CompetitionOpenSinceYear', 'Promo2', 'Promo2SinceWeek',
       'Promo2SinceYear', 'PromoInterval'],
      dtype='object')

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   Store                      1017209 non-null  int64  
 1   DayOfWeek                  1017209 non-null  int64  
 2   Date                       1017209 non-null  object 
 3   Sales                      1017209 non-null  int64  
 4   Customers                  1017209 non-null  int64  
 5   Open                       1017209 non-null  int64  
 6   Promo                      1017209 non-null  int64  
 7   StateHoliday               1017209 non-null  object 
 8   SchoolHoliday              1017209 non-null  int64  
 9   StoreType                  1017209 non-null  object 
 10  Assortment                 1017209 non-null  object 
 11  CompetitionDistance        1014567 non-null  float64
 12  CompetitionOpenSinceMonth  693861 non-null   float64
 13  CompetitionO

In [16]:
df.shape

(1017209, 18)

In [17]:
df.columns.tolist()

['Store',
 'DayOfWeek',
 'Date',
 'Sales',
 'Customers',
 'Open',
 'Promo',
 'StateHoliday',
 'SchoolHoliday',
 'StoreType',
 'Assortment',
 'CompetitionDistance',
 'CompetitionOpenSinceMonth',
 'CompetitionOpenSinceYear',
 'Promo2',
 'Promo2SinceWeek',
 'Promo2SinceYear',
 'PromoInterval']

In [18]:
df.isnull().sum()

Store                             0
DayOfWeek                         0
Date                              0
Sales                             0
Customers                         0
Open                              0
Promo                             0
StateHoliday                      0
SchoolHoliday                     0
StoreType                         0
Assortment                        0
CompetitionDistance            2642
CompetitionOpenSinceMonth    323348
CompetitionOpenSinceYear     323348
Promo2                            0
Promo2SinceWeek              508031
Promo2SinceYear              508031
PromoInterval                508031
dtype: int64

<h3>Filtering closed store days first</h3>

As Closed days have Sales = 0 which will distort every analysis.

In [19]:
print("Rows before filtering closed stores: ", len(df))

Rows before filtering closed stores:  1017209


In [20]:
print("Closed store rows: ", (df["Open"] == 0).sum())

Closed store rows:  172817


In [21]:
df = df[df['Open'] == 1].copy()
print("Rows after filtering:", len(df))

Rows after filtering: 844392


Checking if there are any remaining rows with Sales = 0 even when open?

In [22]:
print("Open but zero sales:", ((df['Open'] == 1) & (df['Sales'] == 0)).sum())

Open but zero sales: 54


<h3>Handling the missing values</h3>

In [23]:
df.isnull().sum()[df.isnull().sum() > 0]

CompetitionDistance            2186
CompetitionOpenSinceMonth    268619
CompetitionOpenSinceYear     268619
Promo2SinceWeek              423307
Promo2SinceYear              423307
PromoInterval                423307
dtype: int64

CompetitionDistance — 2186 nulls<br>
No nearby competitor, so fill with large distance value

In [24]:
df['CompetitionDistance'] = df['CompetitionDistance'].fillna(200000)

CompetitionOpenSinceMonth & Year — 268619 nulls<br>
Store has no competitor, so competition "opened" at time 0

In [25]:
df['CompetitionOpenSinceMonth'] = df['CompetitionOpenSinceMonth'].fillna(0)
df['CompetitionOpenSinceYear'] = df['CompetitionOpenSinceYear'].fillna(0)

Promo2SinceWeek, Promo2SinceYear, PromoInterval — 423307 nulls<br>
Store is not participating in Promo2, so fill with 0 / 'None'

In [26]:
df['Promo2SinceWeek'] = df['Promo2SinceWeek'].fillna(0)
df['Promo2SinceYear'] = df['Promo2SinceYear'].fillna(0)
df['PromoInterval'] = df['PromoInterval'].fillna('None')

In [27]:
df.isnull().sum()

Store                        0
DayOfWeek                    0
Date                         0
Sales                        0
Customers                    0
Open                         0
Promo                        0
StateHoliday                 0
SchoolHoliday                0
StoreType                    0
Assortment                   0
CompetitionDistance          0
CompetitionOpenSinceMonth    0
CompetitionOpenSinceYear     0
Promo2                       0
Promo2SinceWeek              0
Promo2SinceYear              0
PromoInterval                0
dtype: int64

<h3>Fixing data types and extracting date features</h3>

In [28]:
df['Date'] = pd.to_datetime(df['Date'])

In [29]:
df['Year']      = df['Date'].dt.year
df['Month']     = df['Date'].dt.month
df['Week']      = df['Date'].dt.isocalendar().week.astype(int)
df['DayOfWeek'] = df['Date'].dt.dayofweek                                 # 0=Monday, 6=Sunday
df['Quarter']   = df['Date'].dt.quarter

Encoding categorical columns as category dtype

In [30]:
df['StoreType']   = df['StoreType'].astype('category')
df['Assortment']  = df['Assortment'].astype('category')
df['StateHoliday'] = df['StateHoliday'].astype(str)

Verifying the data ranges

In [31]:
print("Date range:", df['Date'].min(), "to", df['Date'].max())
print("Unique years:", df['Year'].unique())

Date range: 2013-01-01 00:00:00 to 2015-07-31 00:00:00
Unique years: [2015 2014 2013]


<h3>Feature engineering</h3>

Log sales — handles right-skewed distribution

In [32]:
df['log_sales'] = np.log1p(df['Sales'])

Competition proximity flag

In [33]:
df['competition_flag'] = (df['CompetitionDistance'] < 1000).astype(int)

Holiday flag — any kind of holiday active

In [34]:
df['is_holiday'] = (
    (df['StateHoliday'] != '0') |
    (df['SchoolHoliday'] == 1)
).astype(int)

Promo2 active flag — is Promo2 currently running this month? <br>
PromoInterval tells you which months Promo2 runs (e.g. "Jan,Apr,Jul,Oct")

In [35]:
month_map = {1:'Jan', 2:'Feb', 3:'Mar', 4:'Apr',
             5:'May', 6:'Jun', 7:'Jul', 8:'Aug',
             9:'Sep', 10:'Oct', 11:'Nov', 12:'Dec'}

In [36]:
def promo2_active(row):
    if row['PromoInterval'] == 'None':
        return 0
    months = row['PromoInterval'].split(',')
    return 1 if month_map.get(row['Month'], '') in months else 0

In [37]:
df['promo2_active'] = df.apply(promo2_active, axis=1)

Weekend flag

In [38]:
# Weekend flag
df['is_weekend'] = (df['DayOfWeek'] >= 5).astype(int)

In [39]:
print("New columns added:")
print(['log_sales', 'competition_flag', 'is_holiday',
       'promo2_active', 'is_weekend'])
print(df[['Sales', 'log_sales', 'Promo', 'promo2_active',
          'is_holiday', 'competition_flag']].head(10))

New columns added:
['log_sales', 'competition_flag', 'is_holiday', 'promo2_active', 'is_weekend']
   Sales  log_sales  Promo  promo2_active  is_holiday  competition_flag
0   5263   8.568646      1              0           1                 0
1   6064   8.710290      1              1           1                 1
2   8314   9.025816      1              1           1                 0
3  13995   9.546527      1              0           1                 1
4   4822   8.481151      1              0           1                 0
5   5651   8.639765      1              0           1                 1
6  15344   9.638545      1              0           1                 0
7   8492   9.046998      1              0           1                 0
8   8565   9.055556      1              0           1                 0
9   7185   8.879890      1              0           1                 0


<h3>Sanity or Data Validation checks before moving on</h3>

Check 1: Sales distribution check — should be right-skewed, no negatives

In [40]:
print("Sales stats:")
print(df['Sales'].describe())
print("Negative sales:", (df['Sales'] < 0).sum())  # Should be 0

Sales stats:
count    844392.000000
mean       6955.514291
std        3104.214680
min           0.000000
25%        4859.000000
50%        6369.000000
75%        8360.000000
max       41551.000000
Name: Sales, dtype: float64
Negative sales: 0


Check 2: Suspicious Stores

It checks for each store, what percentage of days had zero sales?

In [41]:
store_zero_pct = df.groupby('Store').apply(lambda x: (x['Sales'] == 0).mean()).sort_values(ascending=False)
suspicious_stores = store_zero_pct[store_zero_pct > 0.3]

C:\Users\medha\AppData\Local\Temp\ipykernel_9668\578229729.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  store_zero_pct = df.groupby('Store').apply(lambda x: (x['Sales'] == 0).mean()).sort_values(ascending=False)


In [42]:
print(f"\nSuspicious stores (>30% zero sales when open): "
      f"{len(suspicious_stores)}")
print(suspicious_stores.head(10))


Suspicious stores (>30% zero sales when open): 0
Series([], dtype: float64)


Check 3: Promo coverage — what % of open days had a promotion?

In [43]:
print(f"\nPromo coverage: {df['Promo'].mean():.1%}")
print(f"Promo2 active coverage: {df['promo2_active'].mean():.1%}")


Promo coverage: 44.6%
Promo2 active coverage: 16.6%


Check 4: Store count per type

In [44]:
print("\nStores per type:")
print(df.groupby('StoreType')['Store'].nunique())



Stores per type:
StoreType
a    602
b     17
c    148
d    348
Name: Store, dtype: int64


C:\Users\medha\AppData\Local\Temp\ipykernel_9668\975240911.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby('StoreType')['Store'].nunique())


Check 5: Date continuity — are there gaps?

In [45]:
date_range = pd.date_range(df['Date'].min(), df['Date'].max(), freq='D')
actual_dates = df['Date'].unique()
print(f"\nExpected date range length: {len(date_range)}")
print(f"Actual unique dates in data: {len(actual_dates)}")


Expected date range length: 942
Actual unique dates in data: 942


Instead of deleting the suspicious stores, lets just create a flag column

In [46]:
df['suspicious_store'] = df['Store'].isin(suspicious_stores.index).astype(int)

Saving the cleaned dataframe

In [47]:
df.to_csv('../cleaned_data.csv', index=False)

In [48]:
df.shape

(844392, 28)

In [49]:
df.columns

Index(['Store', 'DayOfWeek', 'Date', 'Sales', 'Customers', 'Open', 'Promo',
       'StateHoliday', 'SchoolHoliday', 'StoreType', 'Assortment',
       'CompetitionDistance', 'CompetitionOpenSinceMonth',
       'CompetitionOpenSinceYear', 'Promo2', 'Promo2SinceWeek',
       'Promo2SinceYear', 'PromoInterval', 'Year', 'Month', 'Week', 'Quarter',
       'log_sales', 'competition_flag', 'is_holiday', 'promo2_active',
       'is_weekend', 'suspicious_store'],
      dtype='object')

## Key EDA Findings

1. Promo vs No-Promo raw lift: +38.8% in average daily sales  

2. Store Type B has highest average sales (€10,231), Type D lowest (€6,822).  
   Promotion response likely varies by type.

3. Promotions cluster during seasonally strong periods (December peak, pre-summer). Raw comparison overstates promotion effect.
   This is the core motivation for Difference-in-Differences methodology.

4. Promo rate varies by store type: Type A = 44.7%, Type B = 38.2%, Type C = 44.9%, Type D = 44.7%.  
